### PR TITLE
Vectorize tanh

### PR DESCRIPTION
**Pull request by vedanuj**
This PR adds an AVX vectorized `tanh` implementation and addresses #4984. This is a vectorized version of [cephes math library's single precision](http://www.netlib.org/cephes/) `tanh` function. We see **speedups of upto ~9x**. Benchmark results below.

_Details :_ This revives the old [PR 5987](https://github.com/pytorch/pytorch/pull/5987) which was using an `exp256_ps` vectorized exponentiation method from  `avx_mathfun.h`. But `avx_mathfun` was [imprecise in many cases](https://github.com/pytorch/pytorch/pull/6192). After the addition of [Sleef](https://github.com/pytorch/pytorch/pull/7341) we can now use the `expf` methods from Sleef to replace `exp256_ps` used in the earlier PR. This implementation uses the `Sleef_expf8_u10` function for single precision float. 

**Benchmark results with vectorized `tanh` for float::**

master @ 947155c69ddc114c8f8a69a0f971386fb365bf09

```
['tanh']
tanh:           size: 10^2      count: 10000    elapsed: 8.143314838409424      type: torch.FloatTensor
tanh:           size: 10^3      count: 1000     elapsed: 1.8030304908752441     type: torch.FloatTensor
tanh:           size: 10^4      count: 100      elapsed: 1.4339056015014648     type: torch.FloatTensor
tanh:           size: 10^5      count: 10       elapsed: 1.2880470752716064     type: torch.FloatTensor
```

fa02f2a

```
tanh:           size: 10^2      count: 10000    elapsed: 1.0960958003997803     type: torch.FloatTensor
tanh:           size: 10^3      count: 1000     elapsed: 0.19293761253356934    type: torch.FloatTensor
tanh:           size: 10^4      count: 100      elapsed: 0.5942552089691162     type: torch.FloatTensor
tanh:           size: 10^5      count: 10       elapsed: 0.5254971981048584     type: torch.FloatTensor
```

**Doesn't affect the performance for doubles ::** 

master @ 947155c69ddc114c8f8a69a0f971386fb365bf09

```
['tanh']
tanh:           size: 10^2      count: 10000    elapsed: 9.550215005874634      type: torch.DoubleTensor
tanh:           size: 10^3      count: 1000     elapsed: 2.3597443103790283     type: torch.DoubleTensor
tanh:           size: 10^4      count: 100      elapsed: 1.9869627952575684     type: torch.DoubleTensor
tanh:           size: 10^5      count: 10       elapsed: 1.657977819442749      type: torch.DoubleTensor
```

fa02f2a

```
['tanh']
tanh:           size: 10^2      count: 10000    elapsed: 9.589819431304932      type: torch.DoubleTensor
tanh:           size: 10^3      count: 1000     elapsed: 2.237194299697876      type: torch.DoubleTensor
tanh:           size: 10^4      count: 100      elapsed: 1.8959002494812012     type: torch.DoubleTensor
tanh:           size: 10^5      count: 10       elapsed: 1.7132713794708252     type: torch.DoubleTensor
```

**Correctness Test::**
```
python test_torch.py TestTorch.test_tanh

.
----------------------------------------------------------------------
Ran 1 test in 0.205s
```

@cpuhrsch @apaszke @colesbury Please review

### CODE REVIEW

**Review by apaszke — COMMENTED** [review]
Looks like Sleef already has `Sleef_tanhf8_u10` implemented. Does this implementation has any benefits over that?

**Comment by vedanuj** [comment]
@apaszke  I benchmarked with `Sleef_tanhf8_u10` and it's somewhat slower than this implementation. Sleef also has `avx`/`avx2` versions of these functions which are also slower. 

However if you suggest we use the Sleef version to keep it more consistent thorughout the codebase I can update this PR to use `Sleef_tanhf8_u10`. 

With `Sleef_tanhf8_u10`

```
['tanh']
tanh:           size: 10^2      count: 10000    elapsed: 2.47904634475708       type: torch.FloatTensor
tanh:           size: 10^3      count: 1000     elapsed: 0.5574073791503906     type: torch.FloatTensor
tanh:           size: 10^4      count: 100      elapsed: 0.6806240081787109     type: torch.FloatTensor
tanh:           size: 10^5      count: 10       elapsed: 0.5766274929046631     type: torch.FloatTensor
```

With `Sleef_tanhf8_u10avx`

```
['tanh']
tanh:           size: 10^2      count: 10000    elapsed: 3.0415682792663574     type: torch.FloatTensor
tanh:           size: 10^3      count: 1000     elapsed: 0.740088939666748      type: torch.FloatTensor
tanh:           size: 10^4      count: 100      elapsed: 0.7270100116729736     type: torch.FloatTensor
tanh:           size: 10^5      count: 10       elapsed: 0.5357003211975098     type: torch.FloatTensor

```

With `Sleef_tanhf8_u10avx2`

```
['tanh']
tanh:           size: 10^2      count: 10000    elapsed: 2.3759076595306396     type: torch.FloatTensor
tanh:           size: 10^3      count: 1000     elapsed: 0.6075067520141602     type: torch.FloatTensor
tanh:           size: 10^4      count: 100      elapsed: 0.6829066276550293     type: torch.FloatTensor
tanh:           size: 10^5      count: 10       elapsed: 0.5176279544830322     type: torch.FloatTensor
```

File: aten/src/ATen/cpu/vec256/vec256_double.h [hunk 1]
```diff
@@ -91,6 +91,9 @@ template <> class Vec256<double> {
   Vec256<double> cos() const {
     return map(std::cos);
   }
+  Vec256<double> tanh() const {
+    return map(std::tanh);
```
Comment (cpuhrsch) [inline aten/src/ATen/cpu/vec256/vec256_double.h]: We could use Sleef's tanh for doubles here.

**Comment by cpuhrsch** [comment]
As a friendly suggestion, if your implementation is faster than SLEEF I also suggest reaching out [to them](https://github.com/shibatch/sleef). That could actually have even more impact that just adding this to pytorch.

**Pull request closed without merging**